### PR TITLE
cudatoolkit: add 10.1.243

### PR DESCRIPTION
**Pull request by dguibert**
###### Things done

- [X] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [X] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @knedlsepp @danieldk

### CODE REVIEW

**Comment by danieldk** [comment]
Thanks! I'll try to test this on a non-NixOS system today.

File: pkgs/development/compilers/cudatoolkit/default.nix [hunk 1]
```diff
@@ -304,5 +318,13 @@ in rec {
...
 
+  cudatoolkit_10_1 = common {
+    version = "10.1.243";
+    url = "https://developer.download.nvidia.com/compute/cuda/10.1/Prod/local_installers/cuda_10.1.243_418.87.00_linux.run";
+    sha256 = "0caxhlv2bdq863dfp6wj7nad66ml81vasq2ayf11psvq2b12vhp7";
+
+    gcc = gcc7;
+  };
+
   cudatoolkit_10 = cudatoolkit_10_0;
```
Comment (knedlsepp) [inline pkgs/development/compilers/cudatoolkit/default.nix]: We probably want to default to cudatoolkit_10_1 here.

File: pkgs/top-level/all-packages.nix [hunk 2]
```diff
@@ -2473,7 +2473,8 @@ in
     cudatoolkit_9_1
     cudatoolkit_9_2
     cudatoolkit_10
-    cudatoolkit_10_0;
+    cudatoolkit_10_0
+    cudatoolkit_10_1;
 
   cudatoolkit = cudatoolkit_9;
```
Comment (knedlsepp) [inline pkgs/top-level/all-packages.nix]: Might be time to bump that one to cudatoolkit_10

**Comment by danieldk** [comment]
I couldn't test Tensorflow (I have to recompile it with CUDA 10.1), but CUDA and CuBLAS seem to work well on Ubuntu 18.04 with this PR.

File: pkgs/development/compilers/cudatoolkit/default.nix (hunk 1 above)
Comment (dguibert) [reply pkgs/development/compilers/cudatoolkit/default.nix]: done

File: pkgs/top-level/all-packages.nix (hunk 2 above)
Comment (dguibert) [reply pkgs/top-level/all-packages.nix]: done

**Comment by dguibert** [comment]
> I couldn't test Tensorflow (I have to recompile it with CUDA 10.1), but CUDA and CuBLAS seem to work well on Ubuntu 18.04 with this PR.

I've also update cudnn.

**Review by disassembler — APPROVED** [review]
I approve. If this is ready, lets merge and cherry pick into 19.09. We're trying to close out all open issues in the milestone this week.

**Comment by knedlsepp** [comment]
@disassembler Would you mind merging this?

**Comment by knedlsepp** [comment]
@andir Could you take a look?

**Comment by andir** [comment]
@knedlsepp Looks good. All those now optional build steps are no longer required because they cleaned up their build system?

**Comment by knedlsepp** [comment]
> @knedlsepp Looks good. All those now optional build steps are no longer required because they cleaned up their build system?

Yes, cudatoolkit 10.1 is actually rather straightforward to package. (I can proof that: I did package it independently from this PR starting from a clean slate:  https://github.com/knedlsepp/nixpkgs/commit/630068ecdbe3f53901cee19ddf6a05675ce89766 )

**Pull request merged**